### DISCUSSION TITLE
Python API Nested transactions

### DISCUSSION THREAD

**Discussion by Mulugruntz** [geldata/gel#2426] [Q&A] [answered]
Not sure where to put this post, as it could either be Q&A, Ideas or Show and tell.

I've seen this discussion there https://github.com/edgedb/edgedb/discussions/1708

I came across a moment in my code where a function running queries that had to be in a transaction would not function properly if it was called already inside a transaction.

So I made this decorator:
https://gist.github.com/Mulugruntz/b3fe059477d046904ff997c9e6119719

I'm not sure whether I did right or if nested transactions are indeed supported (but I couldn't make them work).

Quick example:
```python
async def runme(connection):
    async for tx in connection.retrying_transaction():
        async with tx:
            print(await tx.query_one("SELECT 1+1"))
            print(await tx.query_one("SELECT 2+2"))

async def run_client():
    async with db_connect() as connection:
        async for tx in connection.retrying_transaction():
            async with tx:
                await runme(tx)
        await runme(connection)

if __name__ == "__main__":
    asyncio.get_event_loop().run_until_complete(run_client())
```

Here, the `runme` function is sometimes called inside an already existing transaction, sometimes not.
Of course, this does not work and ends up in `AttributeError: 'AsyncIOIteration' object has no attribute 'retrying_transaction'`, implying that nested transactions are not available in the Python driver.

So, is it correct to assume that there is no nested transaction support in EdgeDB?

**Comment by elprans** [comment, marked as answer]
Support for subtransactions is being worked on right now.  See more in the [RFC](https://github.com/edgedb/rfcs/blob/master/text/1004-transactions-api.rst#subtransactions-api).

Subtransactions have to be explicit, because there is no way to retry them: any failure aborts the entire enclosing transaction.

Here's an example on how that'll look like once implemented:

```
async def runme(connection):
    async with connection.subtransaction() as tx:
        print(await tx.query_one("SELECT 1+1"))
        print(await tx.query_one("SELECT 2+2"))

async def run_client():
    async with db_connect() as connection:
        async for tx in connection.retrying_transaction():
            async with tx:
                await runme(tx)
        await runme(connection)

if __name__ == "__main__":
    asyncio.get_event_loop().run_until_complete(run_client())
```

There are good reasons to _not_ implement a seamless composable transaction API where `transaction()` is either a top-level transaction or a sub-transaction:

1. As mentioned above, it is impossible to retry a subtransaction.
2. Transactions can have varied attributes (isolation level, writability, constraint deferrability), and if you nest `transaction()` calls, their attributes must match exactly.

All that said, it's still possible to implement a homogeneous API by combining `raw_transaction()` and `subtransaction()` in a simple wrapper.

**Reply by Mulugruntz** [reply]
Thanks @elprans as always for your quick and detailed answers.

how deep can `.subtransaction()` be nested?

You say that if a sub fails, the whole tree fails (and then we can retry from the top transaction). Then you mention it's *impossible to retry* a transaction. Is there a particular reason to that? Is it more of a design choice? Or bad/good practice?

So far, I could very well see future-me change the decorator to either create a retrying transaction if a pool or connection was given... or creating a subtransaction if a top transaction (retrying / raw) was given.

To me, it would still make sense, even if the subtransaction doesn't retry on its own but if the whole tree it's part of, retries.

**Reply by elprans** [reply]
> how deep can `.subtransaction()` be nested?

There's no practical limit, but there are some performance ramifications of having too many subtransactions (IIRC more than 64).  If you don't need the ability to recover from errors in a subtransaction and continue the enveloping transaction, it is actually advisable to avoid starting a subtransaction altogether (i.e. make it a nop in your wrapper).

> Then you mention it's _impossible to retry_ a transaction. Is there a particular reason to that? Is it more of a design choice? Or bad/good practice?

Some retryable errors, such as a network disconnect, naturally result in transaction being aborted on the server.  `FATAL` Postgres exceptions also result in session termination, but may be retryable.

**Reply by Mulugruntz** [reply]
64 is pretty deep already!
I wouldn't dare imagining a scenario where more than 3-4 would be needed.

So, if I understand correctly, that does mean that retryable transactions (not just raw transactions) will be allowed to have subtransactions. And if it fails anywhere, everything is rolled back and the top retryable transaction will retry.

(sorry for re-asking, maybe it was already obvious, but I just want to make sure I understand properly)

**Reply by elprans** [reply]
Yes, both raw and retryable transactions will be allowed to have subtransactions.

> And if it fails anywhere, everything is rolled back and the top retryable transaction will retry.

If the error is recoverable (e.g. a division by zero), then only the subtransaction will be rolled back and the parent transaction will have the opportunity to handle the error and continue.  If the error is *not* recoverable (e.g. a network disconnect or a FATAL postgres error), then the whole transaction will abort, and if the error is considered to be retryable, the `retryable_transaction` loop will make another attempt.  Currently, only network and transaction serialization errors are considered retryable.

**Reply by Mulugruntz** [reply]
Oohhh good to know that it doesn't retry blindly!

Thanks for the explanation. More complex than I understood at first. So, there's recoverable / non-recoverable and retryable / non-retryable.
